### PR TITLE
Pb 81 lti create new book

### PR DESCRIPTION
**Pull request by bdolor**
WIP 
- clean up documentation
- fix display logic
- improve code coverage

Addresses https://github.com/pressbooks/pressbooks-lti-provider/issues/81

### CODE REVIEW

File: inc/class-tool.php [hunk 1]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+		} else {
+			$i = 1;
+			while ( domain_exists( $parts['host'], $path, 1 ) && $i < 1000 ) {
+				$path = untrailingslashit( $parts['path'] ) . $i;
+				++ $i;
+			}
+		}
+
+		return sprintf( '%1$s://%2$s%3$s', $parts['scheme'], $domain, untrailingslashit( $path ) );
+
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php]: Trailing new line, pls remove

File: inc/class-tool.php [hunk 2]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+			} else {
+				update_blog_option(
+					$book_id, 'pressbooks_lti_consumer_context', [
+						'resource_link_id' => $resource_link_id,
+						'context_id'       => $context_id,
+					]
+				);
+			}
+		} else {
+			$exists = false;
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php]: 'else' block is redundant, we already know that $exists is false

File: tests/test-tool.php [hunk 3]
```diff
@@ -154,4 +194,97 @@ public function test_validateRegistrationRequest() {
...
+
+		$subdomain_happy_path    = 'https://biology201.example.com/';
+		$subdirectory_happy_path = 'https://example.com/biol201';
+		$empty_path              = 'blort@example.com';
+
+		$this->assertEquals( 'https://biology201.example.com', $this->tool->maybeDisambiguateDomain( $subdomain_happy_path ) );
+		$this->assertEquals( 'https://example.com/biol201', $this->tool->maybeDisambiguateDomain( $subdirectory_happy_path ) );
+		$this->assertEmpty( $this->tool->maybeDisambiguateDomain( $empty_path ) );
+		$this->assertEquals( "{$disambiguate}1", $this->tool->maybeDisambiguateDomain( $disambiguate ) );
+
```
Comment (cagp-dev-mtl) [inline tests/test-tool.php:209]: pls remove trailing new lines and comment

File: tests/test-tool.php [hunk 4]
```diff
@@ -94,7 +96,45 @@ public function test_setupUser() {
...
+		$this->tool->setupUser( $consumer_user, $guid2 );
+		$user2 = get_user_by( 'login', $edge_user );
+		$this->assertInstanceOf( '\WP_User', $user2 );
+		$this->assertEquals( $edge_user, $user2->user_login );
+		$this->assertTrue( is_user_member_of_blog( $user2->ID ) );
+
+		// Existing user was recognized and linked
+		$this->assertInstanceOf( '\WP_User', $this->tool->matchUserById( $lti_id2 ) );
+
```
Comment (cagp-dev-mtl) [inline tests/test-tool.php]: Pls remove all trailing newlines before closing brackets

File: inc/class-controller.php [hunk 5]
```diff
@@ -92,6 +92,52 @@ public function contentItemSubmit( $params ) {
...
+	/**
+	 * Creates a book from an LTI request
+	 *
+	 * @param $action
+	 * @param $params
+	 * @since 1.4.0
+	 */
+	public function createBook( $action, $params ) {
+		$connector = Database::getConnector();
+		$tool      = new Tool( $connector );
```
Comment (cagp-dev-mtl) [inline inc/class-controller.php]: Pls keep only one space between the variable and the `=` sign, this applies to all occurrences commited in this PR.

File: inc/class-controller.php [hunk 6]
```diff
@@ -92,6 +92,52 @@ public function contentItemSubmit( $params ) {
...
+			do_exit();
+		}
+
+		if ( empty( $_SESSION['pb_lti_consumer_pk'] ) || empty( $_SESSION['pb_lti_consumer_version'] ) || empty( $_SESSION['pb_lti_return_url'] ) ) {
+			wp_die( __( 'You do not have permission to do that.' ) );
+		}
+
+		$new_book_url = $tool->maybeDisambiguateDomain( $activity_url );
+		$title        = $tool->buildTitle( $_POST['context_label'], $_POST['context_id'], $_POST['resource_link_title'], $_POST['resource_link_id'] );
+
```
Comment (cagp-dev-mtl) [inline inc/class-controller.php]: Trailing newline, pls remove

File: inc/class-tool.php [hunk 7]
```diff
@@ -445,7 +456,7 @@ public function loginUser( $wp_user, $lti_id, $lti_id_was_matched, $role ) {
 	 *
 	 * @return false|\WP_User
 	 */
-	public function matchUser( $lti_id ) {
+	public function matchUserById( $lti_id ) {
 		global $wpdb;
 		$condition = "{$lti_id}|%";
 		$query_result = $wpdb->get_var( $wpdb->prepare( "SELECT user_id FROM {$wpdb->usermeta} WHERE meta_key = %s AND meta_value LIKE %s", self::META_KEY, $condition ) );
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php:462]: Why using a wildcard `LIKE` comparison instead of an exact match?

**Review by cagp-dev-mtl — COMMENTED** [review]
Pls review comments

File: inc/class-controller.php [hunk 8]
```diff
@@ -92,6 +92,52 @@ public function contentItemSubmit( $params ) {
...
+		$lti_id = "{$tool->consumer->consumerGuid}|{$tool->user->getId()}";
+		$wp_user = $tool->matchUserById( $lti_id );
+
+		if ( $wp_user ) {
```
Comment (cagp-dev-mtl) [inline inc/class-controller.php]: two `if` can be flatten in a single one:

`if ( ( $wp_user ) && ( $tool->user->isStaff() || $tool->user->isAdmin() )`

File: inc/class-tool.php [hunk 9]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+	 * Process incoming request, check authenticity of the LTI launch request
+	 *
+	 * @since 1.4.0
+	 *
+	 * @param array $params
+	 */
+	public function processRequest( $params ) {
+		$this->setParams( $params );
+		$this->setParameterConstraint(
+			'oauth_consumer_key', true, 50, [
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php]: `50` is repeated 3 times in this method, can you convert it to a constant?

File: inc/class-tool.php [hunk 10]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+			]
+		);
+		$this->setParameterConstraint( 'resource_link_id', true, 50, [ 'basic-lti-launch-request' ] );
+		$this->setParameterConstraint( 'user_id', true, 50, [ 'basic-lti-launch-request' ] );
+		$this->setParameterConstraint( 'roles', true, null, [ 'basic-lti-launch-request' ] );
+		if ( ! $this->validateRegistrationRequest() ) {
+			$this->ok      = false;
+			$this->message = __( 'Unauthorized registration request. Tool Consumer is not in whitelist of allowed domains.', 'pressbooks-lti-provider' );
+		}
+
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php]: remove trailing newline

File: inc/class-tool.php [hunk 11]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+	public function buildTitle( $course_name, $course_id, $activity_name, $activity_id ) {
+		$course   = ( ! empty( $course_name ) ? $course_name : 'Course ID ' . $course_id );
+		$activity = ( ! empty( $activity_name ) ? $activity_name : 'Activity ID ' . $activity_id );
+
+		$title = sprintf( '%1$s: %2$s', $course, $activity );
+		$title = wp_strip_all_tags( $title, false );
+		$title = ( strlen( $title ) <= 1 ) ? 'Untitled' : $title;
+
+		return $title;
+
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php]: remove trailing newline

File: inc/class-tool.php [hunk 12]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+	 *
+	 * @return int|\WP_Error
+	 * @since 1.4.0
+	 *
+	 */
+	public function createNewBook( $new_book_url, $title, $user_id, $resource_link_id, $context_id ) {
+		$url    = untrailingslashit( $new_book_url );
+		$domain = wp_parse_url( $url, PHP_URL_HOST );
+		$path   = wp_parse_url( $url, PHP_URL_PATH );
+
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php:847]: remove trailing new line

File: inc/class-tool.php [hunk 13]
```diff
@@ -660,4 +671,227 @@ function ( $var ) {
...
+		$book_id = wpmu_create_blog( $domain, $path, $title, $user_id );
+
+		add_blog_option(
+			$book_id, 'pressbooks_lti_consumer_context', [
```
Comment (cagp-dev-mtl) [inline inc/class-tool.php]: there are several occurrences of `pressbooks_lti_consumer_context`, pls convert it to a class constant and if you need it to use it outside of the class, invoke it statically.

**Review by cagp-dev-mtl — COMMENTED** [review]
@bdolor , aside of the comments I left, pls run `composer test` inside vagrant and the plugin root level and ensure all existing tests are passing. These tests are not related to the test coverage decrease but it will let you know if the new code did not introduce regressions.

File: inc/class-tool.php (hunk 13 above)
Comment (bdolor) [reply inc/class-tool.php]: great suggestion, thanks

File: inc/class-tool.php (hunk 9 above)
Comment (bdolor) [reply inc/class-tool.php]: excellent, thank you!

File: inc/class-controller.php (hunk 8 above)
Comment (bdolor) [reply inc/class-controller.php]: complete!

File: inc/class-controller.php (hunk 5 above)
Comment (bdolor) [reply inc/class-controller.php]: done!

File: inc/class-tool.php (hunk 2 above)
Comment (bdolor) [reply inc/class-tool.php]: thank you, good catch.

**Comment by bdolor** [comment]
> @bdolor , aside of the comments I left, pls run `composer test` inside vagrant and the plugin root level and ensure all existing tests are passing. These tests are not related to the test coverage decrease but it will let you know if the new code did not introduce regressions.

no regressions, looking good: 
image

**Comment by cagp-dev-mtl** [comment]
Thanks @bdolor , PR approved and merged, I'll deploy your code now.

**Pull request merged**